tokio-stream/src/stream_ext/peekable.rs

```diff
@@ -0,0 +1,48 @@
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use futures_core::Stream;
+use pin_project_lite::pin_project;
+
+use crate::StreamExt;
+
+pin_project! {
+    /// Stream returned by the [`chain`](super::StreamExt::peekable) method.
+    pub struct Peekable<T: Stream> {
+        peek: Option<T::Item>,
+        #[pin]
+        stream: T,
+    }
+}
+
+impl<T: Stream> Peekable<T> {
+    pub(crate) fn new(stream: T) -> Self {
+        Self { peek: None, stream }
+    }
+
+    /// Peek at the next item in the stream.
+    pub async fn peek(&mut self) -> Option<&T::Item>
+    where
+        T: Unpin,
+    {
+        if let Some(ref it) = self.peek {
+            Some(it)
+        } else {
+            self.peek = self.next().await;
+            self.peek.as_ref()
+        }
```

Comment:
This is only correct if the inner stream is fused. Some usages of `peek` can end up calling `next` on it after it has returned `None`.

Reply:
What should be the behavior in this case? Is it not UB to call next on a stream that has returned `None`?

Reply:
The problem is that even if you _don't_ call `next` after `None` on `Peekable`, it might still happen on the inner stream. For example, if I call `peek` and get `None`, and then call `peek` or `next` afterwards, then that will trigger this case.

You can wrap the inner stream in `Fuse` to avoid this.

(It's not UB to call `next` after `None`, since you can do it without an unsafe block.)

Reply:
I meant UB in a more general sense, as, it's not defined how an iterator must behave after `None`. The  [standard library](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next) says:
> Returns [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None) when iteration is finished. Individual iterator implementations may choose to resume iteration, and so calling next() again may or may not eventually start returning [Some(Item)](https://doc.rust-lang.org/std/option/enum.Option.html#variant.Some) again at some point.

In this case, fusing the iterator would cause unintended effects.

On the other hand [stream](https://docs.rs/futures-core/latest/futures_core/stream/trait.Stream.html#tymethod.poll_next) seems much more opinionated while still reckoning that there is no hard requirement.

I see two options:
- Fuse the inner stream, and document that `Peekable` effectively fuses the stream
- Warn that `peek` may call `next` even after `None` was returned, and one should consider first fusing the stream if that's an issue.

Reply:
Ah, it seems like you can use a double option like how `std::iter::Peekable` is implemented. This way, you have `None` and `Some(None)` to tell the difference between "I'm not storing a peekable value" and "I'm storing a peekable value, and it is `None`".

Reply:
> In this case, fusing the iterator would cause unintended effects.

I don't think it would be a problem to use that solution. After all, it's what the futures crate does.

Reply:
I fused the inner stream 👍